### PR TITLE
Issue/increase awarness of downsides of saving view state

### PR DESCRIPTION
**Pull request by malinajirka**
### Description

The codebase currently contains tens of ViewModels which are storing their complete ViewState into the SavedState bundle by using LiveDataDelegate. However, this approach isn't optimal - [the official documentation even states](https://developer.android.com/topic/libraries/architecture/viewmodel/viewmodel-savedstate) 

`
Note: State must be simple and lightweight. For complex or large data, you should use [local persistence](https://developer.android.com/topic/libraries/architecture/saving-states#local).
`

and

`
Key Point: Usually, data stored in saved instance state is transient state that depends on user input or navigation. Examples of this can be the scroll position of a list, the id of the item the user wants more detail about, the in-progress selection of user preferences, or input in text fields.
`

We are storing way more than that, which is wasting resources, leading to lower performance of the app as well as TransactionTooLarge crashes.

Fixing all these issues is way out of scope of this PR - it's an effort for weeks to implement properly. However, the goal of this PR is to change the API to increases awareness of the downsides of storing more than necessary, especially when creating new VMs.

The approach is deprecating the existing constructor and creating a new factory method which takes care of the creation. I'm intentionally not changing all the existing location to use this factory method as they should be changed only after carefully considering whether we truly store the minimal data.

P.S. I added `unit-tests-exemption` label since 95% of the code of this PR is only about increasing awareness, not actually changing the class.

### Steps to reproduce

No repro steps.

### Testing information

Nothing to test since the production code is unchanged.

- [x] I have considered if this change warrants release notes and have added them to `RELEASE-NOTES.txt` if necessary. Use the "[Internal]" label for non-user-facing changes.

### CODE REVIEW

File: WooCommerce/src/main/kotlin/com/woocommerce/android/viewmodel/LiveDataDelegate.kt [hunk 1]
```diff
@@ -35,7 +54,7 @@ class LiveDataDelegate<T : Parcelable>(
 
     fun observe(owner: LifecycleOwner, observer: (T?, T) -> Unit) {
         if (_liveData.hasActiveObservers()) {
-            throw(IllegalStateException("Multiple observers registered but only one is supported."))
```
Comment (malinajirka) [inline WooCommerce/src/main/kotlin/com/woocommerce/android/viewmodel/LiveDataDelegate.kt:38]: This is unrelated fix for an issue detected by Detekt.

**Comment by dangermattic** [comment]
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Warning">
          2 Warnings
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">This PR is larger than 300 lines of changes. Please consider splitting it into smaller PRs for easier and faster reviews.</td>
    </tr>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">Class <code>LiveDataDelegate</code> is missing tests, but <code>unit-tests-exemption</code> label was set to ignore this.</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_pr-check">
  Generated by :no_entry_sign: <a href="https://danger.systems/">Danger</a>
</p>

**Comment by malinajirka** [comment]
Ouch, I didn't realize we have `no warnings during build` policy, so this approach doesn't actually work 😞. Would you feel comfortable if I added `Suppress` annotation and a comment to all these places? Do you perhaps have another suggestions?

**Comment by hichamboushaba** [comment]
> Ouch, I didn't realize we have `no warnings during build` policy, so this approach doesn't actually work 😞

Yes, that's one of the downsides of such a policy, IMO being able to deprecate stuff temporarily is a good approach for some large changes, but it doesn't work with such policy.

> Would you feel comfortable if I added Suppress annotation and a comment to all these places? Do you perhaps have another suggestions?

I think this would work, the only worry that I have is that with the suppression, those usages will go unnoticed for quite some time, unless we specifically allocate some time to work on them.

PS: I didn't review the whole PR yet, this is just a reply to the last comment, I'm planning to give this some thought before reviewing it.

**Comment by malinajirka** [comment]
> I think this would work, the only worry that I have is that with the suppression, those usages will go unnoticed for quite some time, unless we specifically allocate some time to work on them.

I agree and I'm not necessarily aiming or suggesting we should fix them => my goal is to increases chances no new places (new VMs) using this approach are introduced in the future. Wdyt?

> PS: I didn't review the whole PR yet, this is just a reply to the last comment, I'm planning to give this some thought before reviewing it.

Thanks so much!

File: WooCommerce/src/main/kotlin/com/woocommerce/android/viewmodel/LiveDataDelegate.kt [hunk 2]
```diff
@@ -19,13 +25,26 @@ import kotlin.reflect.KProperty
  *  being updated. If there is more than one, an [IllegalStateException] is thrown.
  *
  */
-class LiveDataDelegate<T : Parcelable>(
-    savedState: SavedStateHandle,
+class LiveDataDelegate<T : Parcelable>
+@Deprecated(
+    message = "Use the provided factory methods or secondary constructors instead.",
+    level = DeprecationLevel.WARNING
+)
```
Comment (hichamboushaba) [inline WooCommerce/src/main/kotlin/com/woocommerce/android/viewmodel/LiveDataDelegate.kt]: I'm just sharing a thought for the discussion here. I understand why are using `Deprecated` here, and it's a very common usage, but technically, we are not deprecating the API. we just want everyone to be aware of its side effects and explicitly say that they need the behavior of saving to the SavedState.

Instead of `Deprecate` here, I was just thinking of using the [`OptIn`](https://kotlinlang.org/docs/opt-in-requirements.html) mechanism, it seems to match what we need here better, and it's the official way for dealing with dangerous APIs (like `GlobalScope` for example), WDYT about using it instead?

I experimented with it, and personally I think it works well, with it there is no need for the factory function, as the constructor itself will require the explicit `OptIn` annotation which should be enough to bring awareness:

```Patch
Index: WooCommerce/src/main/kotlin/com/woocommerce/android/viewmodel/LiveDataDelegate.kt
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/WooCommerce/src/main/kotlin/com/woocommerce/android/viewmodel/LiveDataDelegate.kt b/WooCommerce/src/main/kotlin/com/woocommerce/android/viewmodel/LiveDataDelegate.kt
--- a/WooCommerce/src/main/kotlin/com/woocommerce/android/viewmodel/LiveDataDelegate.kt	(revision fcb79db775b6a8b62025ca733a7e91587ea8f3e0)
+++ b/WooCommerce/src/main/kotlin/com/woocommerce/android/viewmodel/LiveDataDelegate.kt	(date 1721669003564)
@@ -26,10 +26,7 @@
  *
  */
 class LiveDataDelegate<T : Parcelable>
-@Deprecated(
-    message = "Use the provided factory methods or secondary constructors instead.",
-    level = DeprecationLevel.WARNING
-)
+@LiveDelegateSavedStateAPI
 constructor(
     savedState: SavedStateHandle? = null,
     private val initialValue: T,
@@ -37,7 +34,7 @@
     private val onChange: (T?, T) -> Unit = { _, _ -> }
 ) {
     // This is a whitelisted usage of the primary constructor.
-    @Suppress("DEPRECATION")
+    @OptIn(LiveDelegateSavedStateAPI::class)
     constructor(
         initialValue: T,
         onChange: (T?, T) -> Unit = { _, _ -> }
@@ -100,3 +97,11 @@
     @Suppress("DEPRECATION") // This is a whitelisted usage of the primary constructor.
     return LiveDataDelegate(savedState, initialValue, savedStateKey, onChange)
 }
+
+@RequiresOptIn(
+    message = "This API will cause the data to be saved to the SavedState, and should be used with caution.",
+    level = RequiresOptIn.Level.WARNING
+)
+@Retention(AnnotationRetention.BINARY)
+// The name is just a suggestion, if needed, feel free to change it to something more appropriate.
+annotation class LiveDelegateSavedStateAPI
```

What do you think?

**Review by hichamboushaba — APPROVED** [review]
I like the change, nice work, thanks for working on it @malinajirka, I left one thought to discuss, but it's not a blocker.

An additional suggestion: I think we should also add something about this to our internal documentation, we have other APIs that make saving stuff to the SavedState easy, like the extension `SavedStateHandle#getStateFlow`, these APIs are quire useful, so we shouldn't limit their usages, but the developers need to be aware of this limitation, and shouldn't overuse them for saving large instances.

File: WooCommerce/src/main/kotlin/com/woocommerce/android/viewmodel/LiveDataDelegate.kt (hunk 2 above)
Comment (malinajirka) [reply WooCommerce/src/main/kotlin/com/woocommerce/android/viewmodel/LiveDataDelegate.kt]: Thanks for the suggestion @hichamboushaba! I really like it, it's cleaner than the factory method. 

However, I'm not sure what is the best way to ensure others don't end up just copy pasting the usage including the annotation into new VMs without reading what is the annotation really for. That's imo one of the benefits of using deprecation - anyone who copies it sees right away, the usage is not recommended anymore. 

I'm thinking we could perhaps use your approach but instead of opting-in all the existing places, add `@Suppress("OPT_IN_USAGE")` to them with a comment saying something like `Saving more than necessary into the SavedState has associated risks which were not known at the time this field was implemented - after we verify we save only the necessary items, we should replace @Suppress("OPT_IN_USAGE") with @OptIn(LiveDelegateSavedStateAPI::class).`

Does that sound reasonable to you or do you believe just having @OptIn(LiveDelegateSavedStateAPI::class) is all we need?

Comment (malinajirka) [reply WooCommerce/src/main/kotlin/com/woocommerce/android/viewmodel/LiveDataDelegate.kt]: I've added the `Suppress` annotation just to help push things forward - we can now easily use `find and replace` if you believe using `OptIn` without a comment in all the locations is better. Let me know, thanks again @hichamboushaba !

**Comment by malinajirka** [comment]
Thanks so much for the review @hichamboushaba!

> An additional suggestion: I think we should also add something about this to our internal documentation, we have other APIs that make saving stuff to the SavedState easy, like the extension SavedStateHandle#getStateFlow, these APIs are quire useful, so we shouldn't limit their usages, but the developers need to be aware of this limitation, and shouldn't overuse them for saving large instances.

I believe that for something like this any other documentation than in-code documentation won't help tbh. I've added TODOs + comments to the existing places where we use `getStateFlow` and we save the complete ViewState. I think all the changes in this PR should do the trick of increasing awareness of this issue, which is the ultimate goal. Wdyt?

File: WooCommerce/src/main/kotlin/com/woocommerce/android/viewmodel/LiveDataDelegate.kt (hunk 2 above)
Comment (hichamboushaba) [reply WooCommerce/src/main/kotlin/com/woocommerce/android/viewmodel/LiveDataDelegate.kt]: > However, I'm not sure what is the best way to ensure others don't end up just copy pasting the usage including the annotation into new VMs without reading what is the annotation really for. That's imo one of the benefits of using deprecation - anyone who copies it sees right away, the usage is not recommended anymore.

Good point.

> I'm thinking we could perhaps use your approach but instead of opting-in all the existing places, add @Suppress("OPT_IN_USAGE")

I agree, thanks for making this change.

**Comment by hichamboushaba** [comment]
> I believe that for something like this any other documentation than in-code documentation won't help tbh

Thank you @malinajirka for sharing your opinion on this, personally, I believe there is some added value to having these guidelines in the documentation, and some of the reasons that I feel are beneficial are:
1. It helps as a single source of truth and a guideline when someone in a PR review wants to challenge a usage like this and convince the PR author about it.
2. It helps newcomers to our codebase to follow the correct usage.

But this is not a blocker for the PR, so feel free to merge the PR regardless of the decision about this.

**Review by hichamboushaba — APPROVED** [review]
Thanks @malinajirka for the last changes, this looks nice, I hope this helps in reducing the chances of this issue in the future.

I'll leave the decision of when to merge the pull request up to you in case you want to wait for an additional review from someone else.

**Comment by malinajirka** [comment]
> It helps as a single source of truth and a guideline when someone in a PR review wants to challenge a usage like this and convince the PR author about it.
It helps newcomers to our codebase to follow the correct usage.

**Comment by malinajirka** [comment]
Thanks so much for the second review @hichamboushaba !

> It helps as a single source of truth and a guideline when someone in a PR review wants to challenge a usage like this and convince the PR author about it.

> It helps newcomers to our codebase to follow the correct usage.

I'd say that both of these can be achieved with in-code comments - we can link to them and after all the code is the only true source of truth. Creating a separate doc for something like this IMO adds a maintenance overhead and also is technically already documented in [the official Google's documentation](https://developer.android.com/topic/libraries/architecture/viewmodel/viewmodel-savedstate). 

Of course, anyone is welcome to create these docs if they believe it's worth it, I just wanted to share my reasoning why I decided not to work on these.

> I'll leave the decision of when to merge the pull request up to you in case you want to wait for an additional review from someone else.

I'll merge it as soon as I manage to fix the build 🤔  - seems like an unrelated issue not introduced in this PR.

**Comment by hichamboushaba** [comment]
> I'll merge it as soon as I manage to fix the build 🤔 - seems like an unrelated issue not introduced in this PR.

I think UI tests have been flaky lately, and we even [disabled](https://github.com/woocommerce/woocommerce-android/pull/12107) some of them lately, so try to update your branch.
